gateway_code/nodes.py

```diff
@@ -0,0 +1,179 @@
+        """ Setup the control node for the open node autotest """
+        pass
+
+    @abc.abstractmethod
+    def autotest_teardown(self, stop_on):
+        """ Cleanup the control node for the open node autotest """
+        pass
+
+    @abc.abstractmethod
+    def configure_profile(self, profile=None):
+        """ Setup the profile used by the control node """
+        pass
+
+    @abc.abstractmethod
+    def flash(self, firmware_path):
+        """ Flash firmware on the control_node """
+        pass
+
+
+class OpenNode(NodeBase):
+    """ class to inherit, for open node classes """
+    __metaclass__ = MetaNode
+    TYPE = 'undefined'
+
+    @abc.abstractmethod
+    def setup(self, firmware_path):
+        """ Setup the open node with a firmware"""
+        pass
+
+    @abc.abstractmethod
+    def teardown(self):
+        """ Cleanup the open node """
+        pass
+
+
+# import all the nodes/plugins
+def import_all_nodes(pkg_dir):
+    """Looks into the given relative path for modules and imports them"""
+    pkg_dir = os.path.join(os.path.dirname(__file__), pkg_dir)
+    for (module_loader, name, _) in pkgutil.iter_modules([pkg_dir]):
+        module_loader.find_module(name).load_module(name)
+
+
+import_all_nodes('open_nodes')
+import_all_nodes('control_nodes')
+
+
+def _node_class(cld, board_type):
+    """Return the open node class implementation for `board_type`.
+
+    :raises ValueError: if board class can't be found """
+    try:
+        output_class = cld.__registry__[board_type]
+        # Class sanity check
+        assert output_class.TYPE == board_type
+    except KeyError:
+        raise ValueError('Board %s not implemented' % board_type)
+    else:
+        return output_class
+
+
+def _verify_open_node_class(output_class):
+    # Tuple with (class, machine) run 'elftarget.py' on a node firmware
+    assert len(output_class.ELF_TARGET) == 2
+
+    for firmware_attr in ('FW_IDLE', 'FW_AUTOTEST'):
+        firmware = getattr(output_class, firmware_attr, None)
+        assert elftarget.is_compatible_with_node(firmware, output_class), \
+            firmware
+
+    required_autotest = {'echo', 'get_time'}  # mandatory
+    assert required_autotest.issubset(output_class.AUTOTEST_AVAILABLE)
+
+
+def open_node_class(board_type):
+    """Return the open node class implementation for `board_type`.
+
+    :raises ValueError: if board class can't be found """
+    output_class = _node_class(OpenNode, board_type)
+    _verify_open_node_class(output_class)
+    return output_class
+
+
+def control_node_class(cn_type):
+    """Return the control node class implementation for `cn_type`.
+
+    :raises ValueError: if board class can't be found """
+    return _node_class(ControlNode, cn_type)
+
+
+def all_open_nodes_types():
+    """Returns all the open nodes classes"""
+    return OpenNode.__registry__.keys()
+
+
+def all_control_nodes_types():
+    """Returns all the control nodes classes"""
+    return ControlNode.__registry__.keys()
```

Comment:
Just wondering: what happens with nodes that inherit from other nodes (example: ControlNodeIotlabM3 inherits ControlNodeIotlab which itself inherit ControlNode) ?
Does the registry key list contains both or only the direct child (ControlNodeIotlab).

Reply:
inheritance work intuitively I think, if the thing after the `class` keyword defines a `TYPE` attribute, then that type will be used to store in the registry. The only potential trap is a subclass without an explicit TYPE attribute:

```
class BaseOpenNode(OpenNode):
   TYPE = 'base'

class TypeMissingDerivedOpenNode(BaseOpenNode):
   pass

assert open_node_class("base_open_node") == TypeMissingDerivedOpenNode
```


In the case of ControlNodeIotlab, the registry is working as intended (existing unit tests in gateway_code.control_nodes.tests)
```
iotlab
<class 'gateway_code.control_nodes.cn_iotlab.ControlNodeIotlab'>
iotlabm3
<class 'cn_iotlabm3.ControlNodeIotlabm3'>
no
<class 'cn_no.ControlNodeNo'>
```

Reply:
> The only potential trap is a subclass without an explicit TYPE attribute

This exactly what I'm wondering. I'm thinking of factorizing some open nodes code like for example st-lrwan1 and st-iotnode are 2 St-Link based open-nodes, we could make derive from a `NodeStlink(OpenNode)` that would provide TTY, BAUDRATE that are common. The subclass won't have to explicitly provide this and IIUC we are in the trap you mention here.

Reply:
if parent class has no TYPE attribute then it might work though :-) I'm adding a unit test to show what I mean